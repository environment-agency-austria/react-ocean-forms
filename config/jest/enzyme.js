import Enzyme from 'enzyme';
// import Adapter from 'enzyme-adapter-react-16';
import 'babel-polyfill';

// Currently there's no official enzyme support
// for reacts context api. FIXME
// https://github.com/airbnb/enzyme/issues/1553
import Adapter from './enzymeAdapter';

Enzyme.configure({ adapter: new Adapter() });
