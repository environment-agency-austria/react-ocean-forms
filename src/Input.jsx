/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';

import FieldLine from './FieldLine';
import { fieldMetaShape, fieldShape } from './shapes';

/**
 * Component for displaying bootstrap
 * form groups with an html input and
 * oForm support
 */
function Input(props) {
  const {
    field,
    type,
    meta,
  } = props;

  return (
    <FieldLine {...props}>
      {meta.plaintext ? field.value : <input type={type} {...field} />}
    </FieldLine>
  );
}

Input.displayName = 'Input';

Input.defaultProps = {
  type: 'text',
};

Input.propTypes = {
  label: PropTypes.string.isRequired,
  type: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
};

export default Input;
