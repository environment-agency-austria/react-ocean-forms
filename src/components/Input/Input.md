### Examples
#### Input types
Showcase of the input type support

```jsx
import { Form, Input } from 'react-ocean-forms';

function Example() {
  // onChange callback
  const handleChange = (value) => {
    console.log('onChange, value: ' + JSON.stringify(value));
  };

  // onBlur callback
  const handleBlur = () => {
    console.log('onBlur');
  };

  return (
    <Form className="demo">
      <Input
        name="myInput"
        label="Number input"
        onChange={handleChange}
        onBlur={handleBlur}
        type="number"
      />

      <Input
        name="myInput2"
        label="Date input"
        onChange={handleChange}
        onBlur={handleBlur}
        type="date"
      />
    </Form>
  );
}

<Example />
```

#### Custom input
Creating your own input components is quite simple. Note that if you want to use the flexible Intl support, you should put all your text outputs through meta.stringFormatter.

```jsx
import { Form, withField } from 'react-ocean-forms';

/**
 * Simple implementation of a custom input
 */
function BaseCustomInput(props) {
  const {
    field,
    label,
    type,
    meta,
    customProp,
  } = props;


  return (
    <div>
      <div>
        <label htmlFor={field.id}>
          <strong>{meta.stringFormatter(label)}</strong>
        </label>
      </div>
      <input type={type} {...field} />
      {customProp}
    </div>
  );
}
const CustomInput = withField(BaseCustomInput);

function Example() {
  // Submit callback, here you'd make your api calls
  const handleSubmit = (values) => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  // onChange callback
  const handleChange = (value) => {
    console.log('onChange, value: ' + JSON.stringify(value));
  };

  // onBlur callback
  const handleBlur = () => {
    console.log('onBlur');
  };

  return (
    <Form className="demo" onSubmit={handleSubmit}>
      <CustomInput
        name="myCustomInput"
        label="Input with info text"
        onChange={handleChange}
        onBlur={handleBlur}
        customProp="custom text"
      />

      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
    </Form>
  );
}

<Example />
```
