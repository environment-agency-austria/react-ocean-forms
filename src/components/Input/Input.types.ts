/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IFieldComponentProps } from '../Field';

/**
 * Props for the Input component
 */
export interface IInputProps extends IFieldComponentProps {
  /**
   * Input type (e.g. text, number, ...)
   */
  type: string;
}
