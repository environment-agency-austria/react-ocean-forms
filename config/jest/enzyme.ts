import enzyme from 'enzyme';
import enzymeAdapterReact16 from 'enzyme-adapter-react-16';

enzyme.configure({ adapter: new enzymeAdapterReact16() });
