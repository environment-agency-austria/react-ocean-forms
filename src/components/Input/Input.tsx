/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { FieldLine } from '../FieldLine';
import { IInputProps } from './Input.types';

/**
 * Component for displaying bootstrap
 * form groups with an html input and
 * oForm support
 */
export const Input: React.SFC<IInputProps> = (props: IInputProps): JSX.Element => {
  const {
    field,
    type,
    meta,
  } = props;

  const fieldValue = field.value;
  if (typeof fieldValue !== 'string' && fieldValue !== undefined) {
    throw new Error(
      'Incompatible field value supplied for input component '
      + `${field.id}. Only values with type string or undefined are allowed.`,
    );
  }

  return (
    <FieldLine {...props}>
      {meta.plaintext ? field.value : <input type={type} {...field} value={fieldValue} />}
    </FieldLine>
  );
};

Input.displayName = 'Input';

Input.defaultProps = {
  type: 'text',
};
