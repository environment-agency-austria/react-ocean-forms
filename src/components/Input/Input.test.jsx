import React from 'react';
import { shallow } from 'enzyme';

import { Input } from './Input';

describe('<Input />', () => {
  const meta = {
    valid: true,
    error: null,
    isValidating: false,
    touched: false,
    stringFormatter: jest.fn(),
    plaintext: false,
  };
  const field = {
    id: 'unitInput',
    name: 'unitInput',
    value: '',
    disabled: false,
    onChange: jest.fn(),
    onBlur: jest.fn(),
  };

  const setup = props => shallow((
    <Input
      id="unitInput"
      label="unitLabel"
      meta={meta}
      field={field}
      {...props}
    />
  ));

  it('should render without crashing', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should only display the value if plaintext is set', () => {
    meta.plaintext = true;
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
