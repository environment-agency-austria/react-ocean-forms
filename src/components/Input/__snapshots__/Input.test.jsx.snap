// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should only display the value if plaintext is set 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Input
    field={
      Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    id="unitInput"
    label="unitLabel"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      }
    }
    type="text"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": "",
      "field": Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      },
      "id": "unitInput",
      "label": "unitLabel",
      "meta": Object {
        "error": null,
        "isValidating": false,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      },
      "type": "text",
    },
    "ref": null,
    "rendered": "",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": "",
        "field": Object {
          "disabled": false,
          "id": "unitInput",
          "name": "unitInput",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": "",
        },
        "id": "unitInput",
        "label": "unitLabel",
        "meta": Object {
          "error": null,
          "isValidating": false,
          "plaintext": true,
          "stringFormatter": [MockFunction],
          "touched": false,
          "valid": true,
        },
        "type": "text",
      },
      "ref": null,
      "rendered": "",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<Input /> should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Input
    field={
      Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    id="unitInput"
    label="unitLabel"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      }
    }
    type="text"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <input
        disabled={false}
        id="unitInput"
        name="unitInput"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        type="text"
        value=""
      />,
      "field": Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      },
      "id": "unitInput",
      "label": "unitLabel",
      "meta": Object {
        "error": null,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      },
      "type": "text",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "type": "text",
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <input
          disabled={false}
          id="unitInput"
          name="unitInput"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          type="text"
          value=""
        />,
        "field": Object {
          "disabled": false,
          "id": "unitInput",
          "name": "unitInput",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": "",
        },
        "id": "unitInput",
        "label": "unitLabel",
        "meta": Object {
          "error": null,
          "isValidating": false,
          "plaintext": false,
          "stringFormatter": [MockFunction],
          "touched": false,
          "valid": true,
        },
        "type": "text",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "disabled": false,
          "id": "unitInput",
          "name": "unitInput",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
